enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,371 @@
+to see the new plugins.
+
+#### Alternative: ConsolePlugin CRD
+
+Rather than tying plugins to operators, we could have a CRD that allows you to
+declare a plugin.
+
+```yaml
+apiVersion: console.openshift.io/v1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
+  displayName: My Plugin
+  service:
+    name: my-console-plugin
+    namespace: my-operator-namespace
+    port: 8443
+    manifest: /manifest.json
+```
+
+Even if we have the CRD, we'd still want the cluster admin to enable the plugin
+before it becomes active.
+
+A disadvantage to this approach is that it's harder to limit plugins to Red
+Hat operators. It's also harder to trace a plugin back to an operator that
+owns it.
+
+### Loading Plugins
+
+The console backend will proxy requests to the plugin `Service`. This way the
+`Service` itself does not need to be exposed outside the cluster, and avoids
+any same-origin or certificate errors in the browser loading JavaScript from a
+separate domain. 
+
+On load, console will read the manifest for each plugin. The manifest is a JSON
+document that contains metadata about the plugin and the extensions.
+`extensions` mirrors the API used for static plugins, and the same extension
+points will be available.
+
+```json
+{
+  "$schema": "/path/to/schema/manifest.json",
+  "id": "my-plugin",
+  "version": "1.2.3",
+  "pluginAPI": "4.7.x",
+  "displayName": "My Plugin",
+  "extensions": [
+    {
+      "type": "console.flag",
+      "properties": {
+        "handler": {
+          "$codeRef": "example.testHandler"
+        }
+      }
+    },
+    {
+      "type": "console.flag/model",
+      "properties": {
+        "flag": "EXAMPLE",
+        "model": {
+          "group": "example.com",
+          "version": "v1",
+          "kind": "ExampleModel"
+        }
+      }
+    }
+  ]
+}
+```
+
+#### Plugin Manifest Properties
+
+| Property | Description |
+| ----- | -------- |
+| `name: string` | Used as a unique identifier. Each plugin must have a unique name. |
+| `version: string` | The version of the plugin. Version must be parsable by node-semver, eg. “1.2.3” |
+| `pluginAPI: string` | Semver range of compatible OpenShift console versions. |
+| `displayName?: string` | User friendly display name. |
+| `description?: string` | The description of the plugin. |
+| `extensions: Extension[]` | List of extensions contributed by the plugin. |
+| `dependencies?: { [pluginName: string]: string }` | A plugin which depends on other plugins must explicitly define their dependencies. |
+
+##### Extension<P = JSON>
+
+| Property | Description |
+| ----- | -------- |
+| `id: string` | Unique identifier of this extension. |
+| `type: string` | The `type` identifies this extension of a particular extension type. |
+| `properties: P` | The extension properties JSON. |
+| `flags?: Partial<{ required: string[]; disallowed: string[]; }>` | Reference to the flags used to trigger the enablement of this extension. |
```

Comment:
i'm not clear on what this is/how it works.  Wouldn't the plugin just be enabled by referencing its name or something?

Reply:
Yeah, I didn't go into too much detail here because it's something we've already implemented for static plugins. Console has a set of flags it checks during load to detect features on the cluster. This could be the presence of a certain resource, a self subject access review, or some other check contributed by the plugin. Plugins can ask to be disabled if a flag is not set.

I can add more details to the enhancement on this.